testing/scenario_app/ios/Scenarios/Scenarios/AppDelegate.m

```diff
@@ -5,7 +5,7 @@ @implementation AppDelegate
 - (BOOL)application:(UIApplication*)application
     didFinishLaunchingWithOptions:(NSDictionary*)launchOptions {
   self.window = [[UIWindow alloc] initWithFrame:[[UIScreen mainScreen] bounds]];
-  self.window.rootViewController = [[FlutterViewController alloc] init];
+  self.window.rootViewController = [[UIViewController alloc] init];
```

Comment:
I'm not clear on why we want to do this.  Your test should still work without changing this, right?

Reply:
Nope, I have to setup the callback before the view controller is presented otherwise it would be flakey.  We couldn't guarantee that I set the callback up before the first frame got presented.

Reply:
But you're testing a _different_ FlutterViewController, and that one should get its own FirstFrameRendered callback, no?

Reply:
Sure, it may work.  It wasn't necessary for my test or any existing test so I decided to keep it simple.

Reply:
Does the test still pass without changing this?  If so and we can keep it, it makes it a bit easier for people to keep using the actual app to test something.

Reply:
It crashes the test unfortunately:
```
#0	0x00000001012b8d28 in flutter::PersistentCache::ResetStoredNewShaders() [inlined] at /Users/gaaclarke/dev/engine/src/flutter/shell/common/persistent_cache.h:47
#1	0x00000001012b8d28 in flutter::Rasterizer::DoDraw(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >) at /Users/gaaclarke/dev/engine/src/flutter/shell/common/rasterizer.cc:131
#2	0x00000001012b97e4 in decltype(*(std::__1::forward<flutter::Rasterizer*&>(fp0)).*fp(std::__1::forward<std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> > >(fp1))) std::__1::__invoke<void (flutter::Rasterizer::*&)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), flutter::Rasterizer*&, std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >, void>(void (flutter::Rasterizer::*&)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), flutter::Rasterizer*&, std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&) [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/type_traits:4291
#3	0x00000001012b97ac in std::__1::__bind_return<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), std::__1::tuple<flutter::Rasterizer*, std::__1::placeholders::__ph<1> >, std::__1::tuple<std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&>, __is_valid_bind_return<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), std::__1::tuple<flutter::Rasterizer*, std::__1::placeholders::__ph<1> >, std::__1::tuple<std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&> >::value>::type std::__1::__apply_functor<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), std::__1::tuple<flutter::Rasterizer*, std::__1::placeholders::__ph<1> >, 0ul, 1ul, std::__1::tuple<std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&> >(void (flutter::Rasterizer::*&)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), std::__1::tuple<flutter::Rasterizer*, std::__1::placeholders::__ph<1> >&, std::__1::__tuple_indices<0ul, 1ul>, std::__1::tuple<std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&>&&) [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/functional:2226
#4	0x00000001012b97ac in std::__1::__bind_return<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), std::__1::tuple<flutter::Rasterizer*, std::__1::placeholders::__ph<1> >, std::__1::tuple<std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&>, __is_valid_bind_return<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), std::__1::tuple<flutter::Rasterizer*, std::__1::placeholders::__ph<1> >, std::__1::tuple<std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&> >::value>::type std::__1::__bind<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), flutter::Rasterizer*, std::__1::placeholders::__ph<1> const&>::operator()<std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> > >(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&) [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/functional:2259
#5	0x00000001012b97ac in decltype(std::__1::forward<std::__1::__bind<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), flutter::Rasterizer*, std::__1::placeholders::__ph<1> const&>&>(fp)(std::__1::forward<std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> > >(fp0))) std::__1::__invoke<std::__1::__bind<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), flutter::Rasterizer*, std::__1::placeholders::__ph<1> const&>&, std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> > >(std::__1::__bind<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), flutter::Rasterizer*, std::__1::placeholders::__ph<1> const&>&, std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&) [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/type_traits:4350
#6	0x00000001012b97ac in void std::__1::__invoke_void_return_wrapper<void>::__call<std::__1::__bind<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), flutter::Rasterizer*, std::__1::placeholders::__ph<1> const&>&, std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> > >(std::__1::__bind<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), flutter::Rasterizer*, std::__1::placeholders::__ph<1> const&>&, std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&) [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/__functional_base:349
#7	0x00000001012b97ac in std::__1::__function::__func<std::__1::__bind<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), flutter::Rasterizer*, std::__1::placeholders::__ph<1> const&>, std::__1::allocator<std::__1::__bind<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), flutter::Rasterizer*, std::__1::placeholders::__ph<1> const&> >, void (std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >)>::operator()(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&) at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/functional:1572
#8	0x00000001012b8e74 in std::__1::function<void (std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >)>::operator()(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >) const [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/functional:1923
#9	0x00000001012b8e5c in flutter::Pipeline<flutter::LayerTree>::Consume(std::__1::function<void (std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >)>) at /Users/gaaclarke/dev/engine/src/flutter/shell/common/pipeline.h:148
#10	0x00000001012b8b70 in flutter::Rasterizer::Draw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >) at /Users/gaaclarke/dev/engine/src/flutter/shell/common/rasterizer.cc:103
#11	0x00000001012c4e64 in flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_25::operator()() const [inlined] at /Users/gaaclarke/dev/engine/src/flutter/shell/common/shell.cc:794
#12	0x00000001012c4e28 in decltype(std::__1::forward<flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_25&>(fp)()) std::__1::__invoke<flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_25&>(flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_25&) [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/type_traits:4350
#13	0x00000001012c4e28 in void std::__1::__invoke_void_return_wrapper<void>::__call<flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_25&>(flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_25&) [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/__functional_base:349
#14	0x00000001012c4e28 in std::__1::__function::__func<flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_25, std::__1::allocator<flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_25>, void ()>::operator()() at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/functional:1572
#15	0x0000000101274c84 in std::__1::function<void ()>::operator()() const [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/functional:1923
#16	0x0000000101274c70 in fml::MessageLoopImpl::FlushTasks(fml::FlushType) at /Users/gaaclarke/dev/engine/src/flutter/fml/message_loop_impl.cc:138
#17	0x0000000101279e30 in fml::MessageLoopDarwin::OnTimerFire(__CFRunLoopTimer*, fml::MessageLoopDarwin*) at /Users/gaaclarke/dev/engine/src/flutter/fml/platform/darwin/message_loop_darwin.mm:75
#18	0x000000018423baa8 in __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ ()
#19	0x000000018423b76c in __CFRunLoopDoTimer ()
#20	0x000000018423b010 in __CFRunLoopDoTimers ()
#21	0x0000000184238b60 in __CFRunLoopRun ()
#22	0x0000000184158da8 in CFRunLoopRunSpecific ()
#23	0x0000000101279f54 in fml::MessageLoopDarwin::Run() at /Users/gaaclarke/dev/engine/src/flutter/fml/platform/darwin/message_loop_darwin.mm:46
#24	0x0000000101274bc0 in fml::MessageLoopImpl::DoRun() at /Users/gaaclarke/dev/engine/src/flutter/fml/message_loop_impl.cc:84
#25	0x000000010127946c in fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0::operator()() const [inlined] at /Users/gaaclarke/dev/engine/src/flutter/fml/thread.cc:34
#26	0x00000001012793d0 in decltype(std::__1::forward<fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>(fp)()) std::__1::__invoke<fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>(fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0&&) [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/type_traits:4350
#27	0x00000001012793d0 in void std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>&, std::__1::__tuple_indices<>) [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/thread:342
#28	0x00000001012793d0 in void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0> >(void*) at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/thread:352
```

I think running 2 FlutterEngine's is suspect, I've seen some thing that make my scratch my beard looking through the code.

Reply:
I think this change is fine. This delegate is akin to a setup phase in a test. We want it to be as undisturbed as possible so we can observe the side effects of loading vms, isolates, engines, entrypoints etc in the actual tests.

Reply:
At some point we'll want this app to actually run Dart code outside of a test though...  

We should be able to run more than one Flutter engine.  We do it in ios_add2app for example.

Reply:
You know what, that crash I was seeing some times before I added my test.  Now that I've changed the app delegate, I haven't seen it in my test.  It might not be related to having 2 engines, but having to do with running that view controller from the app delegate for some reason.

Reply:
Ahh ok, well we can figure something out for this.  At some point we'll want to be able to present a FlutterViewController automatically in the app but it's fine for now.

Reply:
> At some point we'll want this app to actually run Dart code outside of a test though...

Do we? Can you describe it a bit? I expected everything under testing/... to be just bootstrapping for tests.

Reply:
> We should be able to run more than one Flutter engine. We do it in ios_add2app for example.

We can, but it has side effects (of starting the vm). If we run performance tests against this project, we want to start fresh

Reply:
This isn't a performance test, this is a test that the first frame rendering callback gets called.

See what I'm doing in https://github.com/flutter/engine/pull/10065 - I'd like to do the same for iOS at some point (build the app, upload it to Firebase test lab, get the timeline)

Reply:
Ya I didn't mean this specific test, just the range of tests this project can support (since we're going off of the idea that this is the first useful test we're writing so we're taking over this project). Since many tests can re-use the same project, I was just making the point above that this current form of the project is going to be useful for many tests.

If you then start writing a test checking for cold full-app launch speed, then that test warrants a new project at that moment.